From ddca351e504dc3cb8d91ea10a1863949699c6e50 Mon Sep 17 00:00:00 2001
From: Vita Preskovsky <vitap@ti.com>
Date: Tue, 13 Dec 2011 22:54:47 +0200
Subject: [PATCH] bluez4: fix synchronization between bluetoothd and driver
  * To fix situation when bluetoothd doesn't get messages
    it expects to receive from the kernel. Therefore the
    the required requests are sent again at the stage when device
    is up. The mismatch caused to the list of adapters to be
    empty.

Signed-off-by: Vita Preskovsky <vitap@ti.com>
---
 plugins/hciops.c |   21 +++++++++++++++++++++
 1 files changed, 21 insertions(+), 0 deletions(-)

diff --git a/plugins/hciops.c b/plugins/hciops.c
index ecc0e86..e2198ee 100644
--- a/plugins/hciops.c
+++ b/plugins/hciops.c
@@ -2512,6 +2512,27 @@ static void device_devup_setup(int index)
	hci_send_cmd(dev->sk, OGF_HOST_CTL, OCF_READ_STORED_LINK_KEY,
					READ_STORED_LINK_KEY_CP_SIZE, &cp);

+	/* Even though it shouldn't happen it seems like we might miss
+	 * initialization commands that kernel sends. So we check it here
+	 * to avoid situation when we have device but it is not in the list
+	 * of adapters. */
+       if (hci_test_bit(PENDING_FEATURES, &dev->pending))
+               hci_send_cmd(dev->sk, OGF_INFO_PARAM,
+                                       OCF_READ_LOCAL_FEATURES, 0, NULL);
+
+       if (hci_test_bit(PENDING_VERSION, &dev->pending))
+               hci_send_cmd(dev->sk, OGF_INFO_PARAM,
+                                       OCF_READ_LOCAL_VERSION, 0, NULL);
+
+       if (hci_test_bit(PENDING_NAME, &dev->pending))
+               hci_send_cmd(dev->sk, OGF_HOST_CTL,
+                                       OCF_READ_LOCAL_NAME, 0, NULL);
+
+       if (hci_test_bit(PENDING_BDADDR, &dev->pending))
+               hci_send_cmd(dev->sk, OGF_INFO_PARAM,
+                                       OCF_READ_BD_ADDR, 0, NULL);
+
+
	if (!dev->pending)
		init_adapter(index);
 }
--
1.7.0.4
